OsmAnd/src/net/osmand/plus/plugins/weather/OfflineForecastHelper.java

```diff
@@ -0,0 +1,776 @@
+import net.osmand.core.jni.WeatherTileResourcesManager.DownloadGeoTileRequest;
+import net.osmand.core.jni.WeatherTileResourcesManager.IDownloadGeoTilesAsyncCallback;
+import net.osmand.core.jni.ZoomLevel;
+import net.osmand.map.WorldRegion;
+import net.osmand.plus.OsmandApplication;
+import net.osmand.plus.R;
+import net.osmand.plus.download.DownloadActivityType;
+import net.osmand.plus.download.IndexItem;
+import net.osmand.plus.plugins.PluginsHelper;
+import net.osmand.plus.plugins.weather.indexitem.WeatherIndexItem;
+import net.osmand.plus.settings.backend.OsmandSettings;
+import net.osmand.plus.settings.backend.preferences.CommonPreference;
+import net.osmand.plus.settings.backend.preferences.EnumStringPreference;
+import net.osmand.plus.settings.backend.preferences.ListStringPreference;
+import net.osmand.plus.utils.NativeUtilities;
+import net.osmand.plus.utils.OsmAndFormatter;
+import net.osmand.util.Algorithms;
+
+import org.apache.commons.logging.Log;
+
+import java.math.BigInteger;
+import java.text.DecimalFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class OfflineForecastHelper {
+
+	private static final Log log = PlatformUtil.getLog(OfflineForecastHelper.class);
+
+	private static final String PREF_FORECAST_DOWNLOAD_STATE_PREFIX = "forecast_download_state_";
+	private static final String PREF_FORECAST_LAST_UPDATE_PREFIX = "forecast_last_update_";
+	private static final String PREF_FORECAST_FREQUENCY_PREFIX = "forecast_frequency_";
+	private static final String PREF_FORECAST_TILE_IDS_PREFIX = "forecast_tile_ids_";
+	private static final String PREF_FORECAST_WIFI_PREFIX = "forecast_download_via_wifi_";
+
+	public static final int TILE_SIZE = 40_000;
+	public static final int FORECAST_DATES_COUNT = 24 + (6 * 8) + 1;
+
+	private final OsmandApplication app;
+	private final OsmandSettings settings;
+	private final WeatherCacheSize weatherCacheSize;
+	private WeatherTileResourcesManager weatherResourcesManager;
+	private final Map<String, OfflineForecastInfo> offlineForecastInfo;
+
+	public OfflineForecastHelper(@NonNull OsmandApplication app, @NonNull WeatherCacheSize weatherCacheSize) {
+		this.app = app;
+		settings = app.getSettings();
+		offlineForecastInfo = new HashMap<>();
+		this.weatherCacheSize = weatherCacheSize;
+	}
+
+	public void setWeatherResourcesManager(@NonNull WeatherTileResourcesManager weatherResourcesManager) {
+		this.weatherResourcesManager = weatherResourcesManager;
+	}
+
+	public boolean shouldHaveWeatherForecast(@NonNull WorldRegion region) {
+		String regionId = checkAndGetRegionId(region);
+		int level = region.getLevel();
+
+		boolean russia = RUSSIA_REGION_ID.equals(regionId);
+		boolean russiaPrefix = regionId.startsWith(RUSSIA_REGION_ID);
+		boolean unitedKingdom = regionId.equals(WorldRegion.UNITED_KINGDOM_REGION_ID);
+
+		return isEntireWorld(regionId) ||
```

Comment:
https://github.com/osmandapp/OsmAnd-resources/blob/master/countries-info/regions.xml - boundary - no should be checked